File: 202102220-java-lift-kata/src/test/java/codingdojo/LiftSystemTest.java
package codingdojo;

import org.junit.jupiter.api.Test;

import static java.util.Arrays.asList;
import static org.assertj.core.api.AssertionsForClassTypes.assertThat;

public class LiftSystemTest {

    @Test
    public void lift_should_move_up_to_requested_floor() {
        Lift liftA = new Lift("A", 0, asList(1));
        LiftSystem liftSystem = aLiftSystem(liftA);

        liftSystem.tick();

        Lift lift = liftSystem.getLifts().get(0);
        assertThat(lift.getFloor()).isOne();
    }

    private LiftSystem aLiftSystem(Lift liftA) {
        return new LiftSystem(asList(0, 1, 2), asList(liftA), asList());
    }
}

File: 202102225-kotlin-lift-kata-at-cswc/src/test/kotlin/codingdojo/LiftSystemTest.kt
package codingdojo

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

internal class LiftSystemTest {

    private val liftA = Lift(id = "A", floor = 0, doorsOpen = false)
    private val liftB = Lift(id = "B", floor = 0, doorsOpen = true)
    private val liftSystem = LiftSystem(floors = listOf(0, 1, 2), lifts = listOf(liftA, liftB), calls = emptyList())

    @Test
    fun `should move to requested floor to fulfill a request`() {
        val requestedFloor = 1

        liftA.requestFloor(requestedFloor)
        assertThat(liftA.getFloor()).isNotEqualTo(requestedFloor)
        liftSystem.tick()

        assertThat(liftA.getFloor()).isEqualTo(requestedFloor)
    }

    @Test
    fun `should move to multiple requested floors to fulfill all requests`() {
        liftA.requestFloor(1)
        liftSystem.tick()
        assertThat(liftA.getFloor()).isEqualTo(1)

        liftA.requestFloor(2)
        liftSystem.tick()
        assertThat(liftA.getFloor()).isEqualTo(2)
    }

    @Test
    fun `should not go anywhere if there are not requests`() {
        liftSystem.tick()

        assertThat(liftA.getFloor()).isEqualTo(0)
    }

    @Test
    internal fun `should open doors after fulfilling a request`() {
        liftA.requestFloor(1)

        liftSystem.tick()
        assertThat(liftA.areDoorsOpen()).isFalse()

        liftSystem.tick()
        assertThat(liftA.areDoorsOpen()).isTrue()
    }

    @Test
    internal fun `should close doors after request for floor`() {
        liftB.requestFloor(1)

        liftSystem.tick()

        assertThat(liftB.areDoorsOpen()).isFalse()
    }
}
